Feat/support api key status

xnoname79 opened the pull request.

xnoname79 committed:
feat: support api get search params

xnoname79 committed:
feat: support api key status
support status_code and error_msg for api key
add liberror
add libhttp

TaQuangTu commented on internal/server/server.go:
```diff
@@ -18,84 +21,96 @@ type SearchServer struct {
-			ourLock, err := s.lockDb.LockSimple(ctx, "search:get_key:reset")
-			if err != nil {
-				return
-			}
-			defer ourLock.Unlock()
-			s.keyManager.ResetDailyCounts(100, availableKey.ResetedAt)
-		}(goCtx)
 	}()
```
Could we shorten it to `defer tx.Rollback()`?

xnoname79 replied:
After review the documentation, i noticed it's safe to shorten it since the rollback will be ignored if the tx is committed. So Yes.

TaQuangTu commented on internal/search/key_manager.go:
```diff
@@ -123,6 +116,7 @@ func (m *ApiKeyManager) GetAvailableKey(ctx context.Context, bucketList KeyBucke
         		is_active = TRUE
         		AND id % $1 = $2
         		AND daily_queries > 0
+						AND status_code BETWEEN 200 AND 299
```
Would a key be forgotten forever if it returns failure status code (.e.g 400) once?

xnoname79 replied:
You're right. So I think we don't need this condition of query, we could monitor the error_msg and deactivate those api_keys relate to system-error returned from the API

xnoname79 replied:
Actually adding this more condition is not my prefer, it makes the query seems more confused because the selected key should not acknowledge those statues.

TaQuangTu replied:
Yeah, so let's remove it. And have a retry mechanism in search RPC instead to avoid encountering temporarily broken API keys.

TaQuangTu commented on internal/search/key_manager.go:
```diff
@@ -75,15 +62,19 @@ func (m *ApiKeyManager) GetKeyBucket(ctx context.Context, numberOfPartition int)
 
-func (m *ApiKeyManager) GetAvailableKey(ctx context.Context, bucketList KeyBucketList) (_ *AvailableKey, err error) {
+func (m *ApiKeyManager) GetAvailableKey(ctx context.Context) (_ *AvailableKey, err error) {
+	bucketList, err := m.GetKeyBucket(ctx, 3)
```
Put 3 into a const and better to give it a name, make it more readable for other contributors :smile:

TaQuangTu commented on internal/search/google.go:
```diff
@@ -43,33 +41,18 @@ func (c *Client) Search(ctx context.Context, apiKey, engineID, query string, par
-	if err != nil {
-		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
-	}
+	var (
+		fullURL = fmt.Sprintf("%s?%s", baseURL, queryParams.Encode())
+		result  GoogleSearchResult
+	)
+	status, err = c.httpClient.Request(
+		ctx,
+		libhttp.RequestOption{
+			RequestTimeout: 10 * time.Second,
+			Method:         libhttp.GET,
+			Url:            fullURL,
+		},
+		&result)
```
Love it. Much shorter & clearer

xnoname79 committed:
feat: support curl api
support activate/deactivate keys
support hard delete keys
support add new keys
support get keys

TaQuangTu commented on internal/search/key_manager.go:
```diff
@@ -154,11 +159,107 @@ func (m *ApiKeyManager) ResetDailyCounts(limit int) {
+	for _, row := range data {
+		_, err = stmt.Exec(row...)
+		if err != nil {
+			_ = tx.Rollback()
```
Do we really need rolling back here?

TaQuangTu commented on internal/server/server.go:
```diff
@@ -18,84 +21,165 @@ type SearchServer struct {
+	if searchErr != nil {
+		msg = searchErr.Error()
+	}
+	err = s.keyManager.UpdateKeyStatus(ctx, availableKey.ApiKey, statusCode, msg)
```
For discussion: Should we update the key status in a goroutine (to return search result as quick as possible)?

TaQuangTu commented on config/config.go:
```diff
@@ -10,6 +10,8 @@ type Config struct {
 	REDIS_LOCKDB_DSN       string `env:"REDIS_LOCKDB_DSN"`
 	REDIS_RATE_LIMITDB_DSN string `env:"REDIS_RATE_LIMITDB_DSN"`
+
+	APP_KEY_BUCKET int `env:"APP_KEY_BUCKET"`
```
Better to have a default value, let it be 3.

TaQuangTu commented:
Close because the changes have been included in this [MR](https://github.com/TaQuangTu/findXAI/pull/7)
